Add PAPOTrainer for preference-based optimization

SolarWindRider opened the pull request:
## What does this PR do?

This PR introduces a new trainer named `PAPOTrainer`, which extends `GRPOTrainer` to support the PAPO (Preference Alignment via Policy Optimization) algorithm. 

## Motivation

PAPO is a variant of GRPO that incorporates a contrastive preference optimization mechanism to improve stability when positive samples are sparse. But the official code use verl. To make it convenient for everyone to use, I implemented the TRL version of the code based on the PAPO formula, and it runs successfully.

## Implementation Details

- Added `trl/trainer/papo_trainer.py`
- Added `trl/trainer/papo_config.py`
- Updated `__init__.py` to include `PAPOTrainer`
- All tests pass locally with `pytest tests/trainer/test_papo_trainer.py -v`

## 🧪 Example Usage
https://github.com/SolarWindRider/avr/blob/main/train_papo.py

I have tested my trainer[with PEFT and FSDP] on Ascend910C and H20[single node with 8 cards].

## Checklist
- [x] I have tested this code locally
- [x] I have run all tests with `pytest`
- [x] I have followed the code style guidelines
- [x] I have added docstrings and comments

qgallouedec commented:
Thanks for your contribution. Can you mode this new trainer to trl.experimental instead? Ideally, we would also have a small mention in the paper index section of the documentation

SolarWindRider commented:
> Thanks for your contribution. Can you mode this new trainer to trl.experimental instead? Ideally, we would also have a small mention in the paper index section of the documentation

Thank you for your advice. I have moved this new trainer to trl.experimental and also added PAPO info in paper index.

qgallouedec commented on docs/source/paper_index.md:
```diff
@@ -207,6 +207,14 @@ training_args = GRPOConfig(
+
+### PAPO (Preference Alignment via Policy Optimization)
+
+* **Paper:** [Link to Paper.](https://arxiv.org/abs/2507.06448)
```
can you please try to align with other subsections

````md
### Paper title

**📜 Paper**: https://huggingface.co/papers/XXXX.XXXXX

Some brief intro.

```python
from trl import ...

training_args = ...
```
````

qgallouedec commented on trl/experimental/papo/papo_trainer.py:
```diff
@@ -0,0 +1,380 @@
+        # PAPO doesn't have a published paper yet, so we reference the GRPO paper
+        # and note that PAPO extends it for multimodal reasoning
+        citation = """\
+@article{wang2025perception,
```
can you please remove this method and use a class method for the citation instead. see https://github.com/huggingface/trl/blob/5e691d1bf83ff7080531598b3898adfdd2a790f1/trl/trainer/dpo_trainer.py#L253-L268 for an example

SolarWindRider replied:
done

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_4334). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

SolarWindRider commented:
> LGTM, thanks!

Thank you for your help! 
 {\\__/}
( • - •)
/つThanks
